bluefog/common/operations.cc

```diff
@@ -186,151 +504,401 @@ bool RunLoopOnce(BluefogGlobalState& state) {
         break;
       case MPIOpsType::WIN_ACCUMULATE:
         BFLOG(TRACE, bluefog_global.controller->GetRank())
             << "Processing WIN_ACCUMULATE on " << entry.tensor_name << " with "
             << Vendor_Name(controller_vendor);
-        state.timeline.ActivityStart(entry.tensor_name, "WIN_ACCUMULATE");
+        timeline.ActivityStart(entry.tensor_name, "WIN_ACCUMULATE");
         if (controller_vendor == Vendor::MPI) {
-          state.controller->WinAccumulate(entry);
+          bluefog_global.controller->WinAccumulate(entry);
         }
 #if HAVE_NCCL
         if (controller_vendor == Vendor::NCCL) {
-          state.nccl_controller->WinAccumulate(entry);
+          bluefog_global.nccl_controller->WinAccumulate(entry);
+        }
+#endif
+        timeline.ActivityEnd(entry.tensor_name);
+        break;
+      case MPIOpsType::WIN_CREATE:
+        BFLOG(TRACE, bluefog_global.controller->GetRank())
+            << "Processing WIN_CREATE " << entry.tensor_name << " with "
+            << Vendor_Name(controller_vendor);
+#if HAVE_NCCL
+        if (controller_vendor == Vendor::NCCL) {
+          if (!nccl_context.is_initialized) {
+            bluefog_global.nccl_controller->Initialize();
+            BFLOG(INFO, bluefog_global.controller->GetRank())
+                << "NCCL Initialized";
+          }
+          bluefog_global.nccl_controller->WinCreate(entry);
         }
 #endif
-        state.timeline.ActivityEnd(entry.tensor_name);
+        if (controller_vendor == Vendor::MPI) {
+          bluefog_global.controller->WinCreate(entry);
+        }
+        break;
+      case MPIOpsType::WIN_FREE:
+        BFLOG(TRACE, bluefog_global.controller->GetRank())
+            << "Processing WIN_FREE " << entry.tensor_name << " with "
+            << Vendor_Name(controller_vendor);
+#if HAVE_NCCL
+        // No specified name. So both nccl and mpi will Free win.
+        if (nccl_context.is_initialized && entry.tensor_name.empty()) {
+          bluefog_global.nccl_controller->WinFreeAll(entry);
+        } else {
+          if (controller_vendor == Vendor::NCCL) {
+            bluefog_global.nccl_controller->WinFree(entry);
+          }
+        }
+#endif
+        if (entry.tensor_name.empty()) {
+          bluefog_global.controller->WinFreeAll(entry);
+        } else {
+          if (controller_vendor == Vendor::MPI) {
+            bluefog_global.controller->WinFree(entry);
+          }
+        }
         break;
       default:
-        state.timeline.ActivityEnd(entry.tensor_name);  // End activity for enqueue
+        timeline.ActivityEnd(entry.tensor_name);  // End activity for enqueue
         throw std::runtime_error("Unsupported/Unkown MPI Operation Types");
     }
-  } catch (std::length_error& e) {
-    std::this_thread::sleep_for(std::chrono::microseconds(1));
-  } catch (std::exception& e) {
-    BFLOG(ERROR) << e.what();
   }
-  return !bluefog_global.shut_down;
+}
+
+bool RunLoopOnce(BluefogGlobalState& state) {
```

Comment:
Consider breaking this function as well. 200+ lines here.

Reply:
I prefer to do that under the tensor fusion case. Since it will change the function structure again.